src/rust/session/src/lib.rs

```diff
@@ -182,6 +222,30 @@ impl Session {
     pub fn peer_addr(&self) -> Result<SocketAddr, std::io::Error> {
         self.stream.peer_addr()
     }
+
+    pub fn timestamp(&self) -> Instant {
+        self.timestamp
+    }
+
+    pub fn set_timestamp(&mut self, timestamp: Instant) {
+        self.timestamp = timestamp;
+    }
+
+    pub fn finalize_response(&mut self) {
+        let previous = self.pending_bytes;
+        let current = self.write_pending();
+        let len = current - previous;
+
+        self.pending_bytes = current;
+
+        if len == 0 {
```

Comment:
Does this imply the most recent response is empty (e.g. due to `noreply`)? If so, it'd be nice to add a comment with an example, and why we are recording the request duration here.

Reply:
Yes, comment added